rosbag2_transport/test/rosbag2_transport/test_play_services.cpp

```diff
@@ -0,0 +1,236 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "rosbag2_interfaces/srv/is_paused.hpp"
+#include "rosbag2_interfaces/srv/pause.hpp"
+#include "rosbag2_interfaces/srv/resume.hpp"
+#include "rosbag2_interfaces/srv/toggle_paused.hpp"
+#include "rosbag2_transport/player.hpp"
+#include "test_msgs/msg/basic_types.hpp"
+#include "test_msgs/message_fixtures.hpp"
+
+#include "rosbag2_play_test_fixture.hpp"
+
+using namespace ::testing;  // NOLINT
+
+class PlaySrvsTest : public RosBag2PlayTestFixture
+{
+public:
+  using Pause = rosbag2_interfaces::srv::Pause;
+  using Resume = rosbag2_interfaces::srv::Resume;
+  using TogglePaused = rosbag2_interfaces::srv::TogglePaused;
+  using IsPaused = rosbag2_interfaces::srv::IsPaused;
+
+  PlaySrvsTest()
+  : RosBag2PlayTestFixture(),
+    client_node_(std::make_shared<rclcpp::Node>("test_play_client"))
+  {}
+
+  ~PlaySrvsTest() override
+  {
+    exec_.cancel();
+    rclcpp::shutdown();
+    spin_thread_.join();
+    play_thread_.join();
```

Comment:
For the sake of defensive programming I would recommend to check if threads still running before calling join otherwise it will cause program termination.
i.e.
```cpp
if (spin_thread_.joinable()) {
  spin_thread_.join();
}
if (play_thread_.joinable()) {
  play_thread_.join();
}
```

Reply:
> before calling join otherwise it will cause program termination

That is not true according to  https://en.cppreference.com/w/cpp/thread/thread/join - which states that 

> Blocks the current thread until the thread identified by *this finishes its execution.

This is the intention - the threads will exit after `rclcpp::shutdown` - we want to wait until they do.

Reply:
I meant the case when thread have already finished for some unexpected reasons or haven't even had a chance to start. e.g. caught some exception before calling
```cpp
spin_thread_ = std::thread(
      [this]() {
        exec_.spin();
      });
```
But never mind, this is just in tests. It will not affect quality of the core rosbag2 code. \
Sorry for attracting attention about it, I missed that it's just in tests.